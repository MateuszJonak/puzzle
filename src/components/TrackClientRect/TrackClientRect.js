import React, { Component } from 'react';
import { connect } from 'react-redux';
import { findDOMNode } from 'react-dom';
import PropTypes from 'prop-types';
import actions from '../../store/ui/actions';

// TODO Do some tests
export class TrackClientRect extends Component {
  constructor(props) {
    super(props);

    this.ref = React.createRef();
  }

  componentDidMount() {
    const clientRect = this.getRefBoundingClientRect();
    this.props.rectSet({ name: this.props.name, clientRect });
  }

  getRefBoundingClientRect() {
    return findDOMNode(this.ref.current).getBoundingClientRect();
  }

  render() {
    const { as: Component, children, rectSet, name, ...rest } = this.props;
    return (
      <Component {...rest} ref={this.ref}>
        {children}
      </Component>
    );
  }
}

const mapStateToProps = () => ({});
const mapDispatchToProps = { rectSet: actions.rect.set };

TrackClientRect.defaultProps = {
  as: 'div',
  rectSet: () => {},
};

TrackClientRect.propTypes = {
  name: PropTypes.string.isRequired,
  as: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  rectSet: PropTypes.func,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(TrackClientRect);
