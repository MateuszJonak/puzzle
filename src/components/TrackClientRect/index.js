export { default as TrackClientRect } from './TrackClientRect';
