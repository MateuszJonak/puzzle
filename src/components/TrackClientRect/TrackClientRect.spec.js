import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { TrackClientRect } from './TrackClientRect';

const mockGetBoundingClientRect = jest.fn();
jest.mock('react-dom', () => ({
  findDOMNode: jest.fn().mockImplementation(() => ({
    getBoundingClientRect: mockGetBoundingClientRect,
  })),
}));

describe('TrackClientRect', () => {
  let props, tree, rectSetMock;
  const shallowEnzyme = props => shallow(<TrackClientRect {...props} />);
  const mountEnzyme = props => mount(<TrackClientRect {...props} />);

  beforeEach(() => {
    rectSetMock = jest.fn();
    props = {
      name: 'rect1',
      as: 'div',
      rectSet: rectSetMock,
    };
    tree = shallowEnzyme(props);
  });

  it('should have instance of GameController', () => {
    const instance = tree.instance();

    expect(instance).toBeInstanceOf(TrackClientRect);
  });

  it('should render component correctly with props', () => {
    tree.setProps({
      children: <h1>title</h1>,
    });
    expect(toJson(tree)).toMatchSnapshot();
  });

  describe('Instance tests', () => {
    let instance;

    beforeEach(() => {
      instance = tree.instance();
    });

    describe('getRefBoundingClientRect', () => {
      const resultGetBoundingClientRect = 'boundingClientRect';
      beforeEach(() => {
        instance.ref = {
          current: 'current',
        };
        mockGetBoundingClientRect.mockReturnValue(resultGetBoundingClientRect);
      });

      it('should return bounding client rect', () => {
        expect(instance.getRefBoundingClientRect()).toBe(
          resultGetBoundingClientRect,
        );
      });
    });
  });

  describe('Component did mount', () => {
    let getRefBoundingClientRectSpyOn;
    const resultGetBoundingClientRect = 'boundingClientRect';

    beforeEach(() => {
      getRefBoundingClientRectSpyOn = jest.spyOn(
        TrackClientRect.prototype,
        'getRefBoundingClientRect',
      );
      getRefBoundingClientRectSpyOn.mockReturnValue(
        resultGetBoundingClientRect,
      );
    });

    it('should have called rectSet when game is running', () => {
      tree = mountEnzyme(props);
      expect(getRefBoundingClientRectSpyOn).toHaveBeenCalled();
      expect(rectSetMock).toHaveBeenCalledWith({
        name: props.name,
        clientRect: resultGetBoundingClientRect,
      });
    });
    afterEach(() => {
      getRefBoundingClientRectSpyOn.mockClear();
    });
  });
});
