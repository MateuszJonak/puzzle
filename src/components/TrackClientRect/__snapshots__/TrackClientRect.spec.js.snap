// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackClientRect should render component correctly with props 1`] = `
<div>
  <h1>
    title
  </h1>
</div>
`;
